backend/app/spread/tests/test_update_balances_with_spread.py

```diff
@@ -0,0 +1,126 @@
+import dateutil.parser
+import pandas as pd
+import pytest
+from dateutil.tz import UTC
+
+from backend.app.spread.actions import update_balances_with_spread
+from backend.app.treasury import Balances
+
+
+def _make_balance_series(balances: list[float], strdates: list[str]) -> pd.Series:
+    return pd.Series(
+        balances,
+        index=pd.Index(
+            [
+                dateutil.parser.parse(strdate).replace(tzinfo=UTC)
+                for strdate in strdates
+            ],
+            name="timestamp",
+        ),
+    )
+
+
+def test_token_to_0_percent_usdc():
+    balances = Balances(
+        balances={"GVT": _make_balance_series([123, 246], ["2022-01-01", "2022-01-02"])}
+    )
+    spread_balances = update_balances_with_spread(
+        balances, "GVT", "USDC", 0, "2022-01-01", "2022-01-02"
+    )
+    assert "GVT" in spread_balances.balances
+    assert "USDC" in spread_balances.balances
+    assert (spread_balances.balances["GVT"] + spread_balances.balances["USDC"]).loc[
+        "2022-01-01"
+    ] == 123
+    assert spread_balances.balances["USDC"].tolist() == [0, 0]
+
+
+def test_token_to_20_percent_usdc():
+    balances = Balances(
+        balances={"GVT": _make_balance_series([123, 246], ["2022-01-01", "2022-01-02"])}
+    )
+    spread_balances = update_balances_with_spread(
+        balances, "GVT", "USDC", 20, "2022-01-01", "2022-01-02"
+    )
+    assert "GVT" in spread_balances.balances
+    assert "USDC" in spread_balances.balances
+    assert (
+        spread_balances.balances["GVT"].loc["2022-01-01"]
+        + spread_balances.balances["USDC"].loc["2022-01-01"]
+        == 123
+    )
+    assert spread_balances.balances["USDC"].loc["2022-01-01"] == pytest.approx(
+        123 * 20 / 100.0
+    )
+
+
+def test_token_to_usdc_earlier_start():
+    "Test correct behavior when start date is before any first balance value"
+    balances = Balances(
+        balances={"GVT": _make_balance_series([123, 246], ["2022-01-02", "2022-01-03"])}
+    )
+    spread_balances = update_balances_with_spread(
+        balances, "GVT", "USDC", 20, "2022-01-01", "2022-01-03"
+    )
+    assert "GVT" in spread_balances.balances
+    assert "USDC" in spread_balances.balances
+    assert spread_balances.balances["USDC"].loc["2022-01-01"] == 0
```

Comment:
What would the values be on "2022-01-02" - "2022-01-03"?

Showing this would probably tell more about the expected behaviour of `update_balances_with_spread`

Reply:
That's not an easy question, hence I haven't described this in the tests :)  
Here I'll answer to the general question, not specifically to the case tested in `test_token_to_usdc_earlier_start`, which is a trivial specific case.

We perform the _backtested swap_ (meaning, the balance change which consists of swapping the _token to divest from_, here GVT, with the _spread token_, USDC) at _start date_, here 2022-01-01. So what is kept invariant by the swap on this date is clear: the swap doesn't change the total balance (the sum of original balances on that date). The code is testing this in every case.

Then for a later date, the invariant is a bit harder to define. Really what goes on is that the proportion GVT/USDC is moving for example from 100%/0% to 80%/20% (that's the case tested by `test_token_to_20_percent_usdc`). For dates > _start date_, this ratio is all we care about, we don't need to assert anything about the balance values themselves.

I'll add assertions that check that the ratio is as expected on each date > _start date_.

Reply:
I updated tests. Erratum in what I said above: the ratio is not kept for dates > _start dates_. I tested the invariance of rates of return.